cachito/workers/pkg_managers/golang.py

```diff
@@ -123,10 +131,150 @@ def resolve_gomod_deps(app_source_path, request_id, dep_replacements=None):
+        versioned commit.
+    :param int module_major_version: the Go module's major version as stated in its go.mod file. If
+        this and semantic_version are not provided, 0 is assumed.
+    :return: the Go module's pseudo-version as returned by `go list`
+    :rtype: str
+    """
+    # Use this instead of commit.committed_datetime so that the datetime object is UTC
+    committed_dt = datetime.utcfromtimestamp(commit.committed_date)
+    commit_timestamp = committed_dt.strftime(r'%Y%m%d%H%M%S')
+    commit_hash = commit.hexsha[0:12]
+
+    # vX.0.0-yyyymmddhhmmss-abcdefabcdef is used when there is no earlier versioned commit with an
+    # appropriate major version before the target commit
+    if semantic_version is None:
+        # If the major version isn't in the import path and there is not a versioned commit with the
+        # version of 1, the major version defaults to 0.
+        return f'v{module_major_version or "0"}.0.0-{commit_timestamp}-{commit_hash}'
+
+    # An example of a semantic version with a prerelease is v2.2.0-alpha
+    if semantic_version.prerelease:
+        # vX.Y.Z-pre.0.yyyymmddhhmmss-abcdefabcdef is used when the most recent versioned commit
+        # before the target commit is vX.Y.Z-pre
+        version_seperator = '.'
+        pseudo_semantic_version = semantic_version
+    else:
+        # vX.Y.(Z+1)-0.yyyymmddhhmmss-abcdefabcdef is used when the most recent versioned commit
+        # before the target commit is vX.Y.Z
+        version_seperator = '-'
+        pseudo_semantic_version = semver.bump_patch(str(semantic_version))
+
+    return f'v{pseudo_semantic_version}{version_seperator}0.{commit_timestamp}-{commit_hash}'
+
+
+def get_golang_version(module_name, git_path, target_ref, update_tags=False):
+    """
+    Get the version of the Go module in the input Git repository in the same format as `go list`.
+
+    If target_ref doesn't point to a commit with a semantically versioned tag, a pseudo-version
+    will be returned.
+
+    :param str module_name: the Go module's name
+    :param str git_path: the path to the Git repository
+    :param str target_ref: the Git reference of the Go module to get the version for
+    :param bool update_tags: determines if `git fetch --tags --force` should be run before
+        determining the version
+    :return: a version as `go list` would provide
+    :rtype: str
+    """
+    # If the module is version v2 or higher, the major version of the module is included as /vN at
+    # the end of the module path. If the module is version v0 or v1, the major version is omitted
+    # from the module path.
+    module_major_version = None
+    match = re.match(r'(?:.+/v)(?P<major_version>\d+)$', module_name)
+    if match:
+        module_major_version = int(match.groupdict()['major_version'])
+
+    repo = git.Repo(git_path)
+    if update_tags:
+        repo.remote().fetch(force=True, tags=True)
```

Comment:
Something to think about in the future: we may want to treat failure to update the tags due to network issues as a warning, and allow the rest to proceed. Otherwise, we lose the offline capability for things we have already processed. (I do understand that we should always _try_ to update the tags.)

Reply:
Good point. I'll wrap this in a try/except.

Reply:
Good point. I'll add a try/except.